### PR TITLE
MM-16480 Don't store incomplete channels 

### PR DESCRIPTION
**Pull request by hmhealey**
This bug was happening because users receive `channel_deleted` events for channels that they're not members of, and that caused the reducer for `entities.channels.channels` to an incomplete channel object that only contains a `delete_at` timestamp and nothing else. When passed into `sortChannelsByDisplayName`, this caused it to throw an error because the channel's `name` field was missing.

This fixes the issue by preventing the channels reducer from storing incomplete channels. I also added some additional protection to `getMyFirstChannelForTeams` since it seems to be the only place where we pass channels into `sortChannelsByDisplayName` without indirectly checking if the channel is complete by seeing things like if the current user is a member of it.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-16480

#### Checklist
- Added or updated unit tests (required for all new features)

### CODE REVIEW

File: src/selectors/entities/channels.js [hunk 1]
```diff
@@ -1209,15 +1209,25 @@ export const getDefaultChannelForTeams: (GlobalState) => RelationOneToOne<Team,
...
-    (allChannels: IDMappedObjects<Channel>, myChannelMemberships: RelationOneToOne<Channel, ChannelMembership>, currentUser: UserProfile): RelationOneToOne<Team, Channel> => {
+    (allChannels: IDMappedObjects<Channel>, myChannelMemberships: RelationOneToOne<Channel, ChannelMembership>, myTeams: Array<Team>, currentUser: UserProfile): RelationOneToOne<Team, Channel> => {
         const locale = currentUser.locale || General.DEFAULT_LOCALE;
+
         const result = {};
-        for (const channel of Object.keys(allChannels).map((key) => allChannels[key]).sort(sortChannelsByDisplayName.bind(null, locale))) {
-            if (!result[channel.team_id] && myChannelMemberships[channel.id]) {
-                result[channel.team_id] = channel;
+        for (const team of myTeams) {
+            // Get a sorted array of all channels in the team that the current user is a member of
```
Comment (lieut-data) [inline src/selectors/entities/channels.js]: Appreciate the extra guard, but wondering if it's strictly necessary? If we added all checks downstream like this, the code could quickly become cluttered.

Comment (hmhealey) [reply src/selectors/entities/channels.js]: It's not strictly necessary, but I don't think it really adds complexity in this case since I'm kind of just rewriting the same function in a different order by checking channel membership before sorting. There is a small functional change since we won't return channels for teams that you're not a member of, but I would've expected the function to work this way anyway.

File: src/actions/channels.test.js [hunk 2]
```diff
@@ -1550,7 +1550,7 @@ describe('Actions.Channels', () => {
 
     it('updateChannelHeader', async () => {
         nock(Client4.getChannelsRoute()).
-            post(`${TestHelper.basicChannel.id}`).
+            get(`/${TestHelper.basicChannel.id}`).
```
Comment (jespino) [inline src/actions/channels.test.js]: Why this change?

File: src/actions/channels.test.js [hunk 3]
```diff
@@ -1567,7 +1567,7 @@ describe('Actions.Channels', () => {
 
     it('updateChannelPurpose', async () => {
         nock(Client4.getChannelsRoute()).
-            post(`${TestHelper.basicChannel.id}`).
+            get(`/${TestHelper.basicChannel.id}`).
```
Comment (jespino) [inline src/actions/channels.test.js]: and this one

File: src/actions/channels.test.js (hunk 2 above)
Comment (hmhealey) [reply src/actions/channels.test.js]: The getChannel API wasn't mocked properly before since it uses a GET request, not a POST. The test still passed because of the bug that this PR fixes, so we had to fix the mock to be able to fix the bug.

File: src/actions/channels.test.js (hunk 3 above)
Comment (hmhealey) [reply src/actions/channels.test.js]: Same here
